akka-stream/src/main/scala/akka/stream/javadsl/Sink.scala

```diff
@@ -285,6 +285,15 @@ object Sink {
    * exposes [[ActorMaterializer]] which is going to be used during materialization and
    * [[Attributes]] of the [[Sink]] returned by this method.
    */
+  def onMaterialization[T, M](factory: BiFunction[Materializer, Attributes, Sink[T, M]]): Sink[T, CompletionStage[M]] =
+    scaladsl.Sink.onMaterialization((mat, attr) => factory(mat, attr).asScala).mapMaterializedValue(_.toJava).asJava
+
+  /**
+   * Defers the creation of a [[Sink]] until materialization. The `factory` function
+   * exposes [[ActorMaterializer]] which is going to be used during materialization and
+   * [[Attributes]] of the [[Sink]] returned by this method.
+   */
+  @deprecated("Use 'onMaterialization' instead", "2.6.0")
   def setup[T, M](factory: BiFunction[ActorMaterializer, Attributes, Sink[T, M]]): Sink[T, CompletionStage[M]] =
```

Comment:
This was added in https://github.com/akka/akka/pull/26477 and backported to 2.5, so it's right that we have to deprecate and use a different name.

Not sure `onMaterialization` is a good name though? Sounds like a callback, although the primary purpose is to be a factory?

`setupOnMaterialization` ?

Reply:
I discussed it with Enno and we agreed `setup` isn't a great name to start with. I think `onMaterialization` pretty accurately describes what it does, but I agree it sounds a bit like a callback. Should it perhaps borrow from the `flatMap` terminology since it takes the output stream and "flattens" into the stream?

Reply:
`flatMap` isn't well known in Java, and what are you suggesting? `onMaterializationFlatMap`?

Reply:
`fromMaterializer`?

Reply:
`fromMaterializer` is perfect